skglm/solvers/group_bcd_solver.py

```diff
@@ -0,0 +1,121 @@
+import numpy as np
+from numba import njit
+
+
+def bcd_solver(X, y, datafit, penalty, w_init=None,
+               max_iter=1000, max_epochs=100, tol=1e-7, verbose=False):
+    """Run a group BCD solver.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        Design matrix.
+
+    y : array, shape (n_samples,)
+        Target vector.
+
+    datafit : instance of BaseDatafit
+        Datafit object.
+
+    penalty : instance of BasePenalty
+        Penalty object.
+
+    w_init : array, shape (n_features,), default None
+        Initial value of coefficients.
+        If set to None, a zero vector is used instead.
+
+    max_iter : int, default 1000
+        Maximum number of iterations.
+
+    max_epochs : int, default 100
+        Maximum number of epochs.
+
+    tol : float, default 1e-6
+        Tolerance for convergence.
+
+    verbose : bool, default False
+        Amount of verbosity. 0/False is silent.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Solution that minimizes the problem defined by datafit and penalty.
+
+    p_objs_out: array (max_iter,)
+        The objective values at every outer iteration.
+
+    stop_crit: float
+        The value of the stop criterion.
+    """
+    n_features = X.shape[1]
+    n_groups = len(penalty.grp_ptr) - 1
+
+    # init
+    w = np.zeros(n_features) if w_init is None else w_init
+    Xw = X @ w
+    datafit.initialize(X, y)
+    all_groups = np.arange(n_groups)
+    p_objs_out = np.zeros(max_iter)
+
+    for t in range(max_iter):
+        if t == 0:  # avoid computing p_obj twice
+            prev_p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+
+        for epoch in range(max_epochs):
+            _bcd_epoch(X, y, w, Xw, datafit, penalty, all_groups)
+
+            if epoch % 10 == 0:
+                current_p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+                stop_crit_in = prev_p_obj - current_p_obj
+
+                if max(verbose - 1, 0):
+                    print(
+                        f"Epoch {epoch+1}: {current_p_obj:.10f}"
+                        f"stopping crit: {stop_crit_in:.2f}"
+                    )
+
+                if stop_crit_in <= 0.3 * tol:
```

Comment:
if you're in the inner loop, the criterion is to stop if stop_crit_in is lower than tol, or 0.3 * stop_crit_out

If you're in the outer loop, the criterion is stop_crit(_out) < tol


The current design is inefficient. Once you reach tol * 0.3, you'll exit from the inner loop, but not from the outer, and you'll do that a lot of time.

Replace by <= tol here since you're not using working sets yet anyway.